contribute-to-tidb/code-style-and-quality-guide.md

```diff
@@ -38,3 +38,19 @@ func main() {
     fmt.Println(ping.Domain())
 }
 ```
+
+## Consider using values instead of pointers
+
+Many Go programmers over-use pointers as opposed to just using values directly. Go programs often return pointers to structs with a wrapper and then all functions use the struct as a pointer:
+
+``` go
+type struct S {}; func NewStruct() *S
+
+func (s *S) structMethod()
+```
+
+Using pointers for the entire method set means we can no longer differentiate by the method signature when a function mutates the struct and when it simply reads data. This increases the chances of errors due to the struct ending up in an unintended state.
+
+Consider whether you can use the struct directly without a pointer. The main reason not to do this is to be able to meet the requirements of an interface.
+
+One might assume that pointers would always perform better. However, this is only the case for methods that are called frequently (generally in a for loop) or if the struct is very large. For infrequently accessed code and smaller structs, copying structs can actually perform significantly better because they avoid any GC overhead.
```

Comment:
Is there any empirical value of the struct size that a value receiver is better than a pointer receiver? This would help the audience to make a decision about whether to use pointer or value receivers.

Reply:
That's a good question. Unfortunately the answer is: it depends on the program. The way I look at it though is that if you want predictable performance (good for latency) then you should heavily favor values. Pointers mean the potential for GC which often means slower performance, but even when it means faster throughput it could still mean less predictable performance. Some of this is [discussed well in this post](https://segment.com/blog/allocation-efficiency-in-high-performance-go-services/)

There are already lots of pointers in Go (string, map, array) so for me it usually seems hard to build up large structs unless there is some nesting of structs. So one rule of thumb might be to always consider using a pointer for a struct member that itself is a struct with multiple members.

Reply:
there are some discussions which we can take into account:

- [GO: Method Receiver - Pointer v/s Value](https://medium.com/globant/go-method-receiver-pointer-vs-value-ffc5ab7acdb)
- [Go Best Practices: Pointer or value receivers?](https://flaviocopes.com/golang-methods-receivers/)
- [Value receiver vs. pointer receiver](https://stackoverflow.com/questions/27775376/value-receiver-vs-pointer-receiver)

Reply:
What from those discussions are helpful for you? One thing was that it points out concurrency safety (which is of course a benefit of being read-only). I found one article with a benchmark that found that [performance differed after 9 words](https://h12.io/article/value-vs-pointer-receivers). But I still think it depends on how your program differs from such micro benchmarks.